src/box/memtx_engine.cc

```diff
@@ -483,9 +556,20 @@ memtx_engine_begin_final_recovery(struct engine *engine)
 	if (memtx->state == MEMTX_OK)
 		return 0;
 
+	/* Build the PK of the last recovered space. */
 	assert(memtx->state == MEMTX_INITIAL_RECOVERY);
-	/* End of the fast path: loaded the primary key. */
-	space_foreach(memtx_end_build_primary_key, memtx);
+	struct space *last_recovered_space =
+		space_by_id(memtx->recovery.last_space_id);
+	VERIFY(memtx_end_build_primary_key(last_recovered_space, memtx) == 0);
+
+	/**
+	 * There may be spaces we had prepared a fast PK build path for, but
+	 * they had no inserts performed during the snapshot recovery, so no
+	 * PK build had been started nor finished for them. Let's just update
+	 * the replace callback for such spaces, so they have the same one as
+	 * all other spaces.
+	 */
+	space_foreach(memtx_begin_empty_space_final_recovery, memtx);
```

Comment:
Wouldn't it be better to call this function for each space (both empty and not)? AFAICS it's required solely to set the replace callback. I think we can postpone it for non-empty spaces until final recovery completion as well.